service/matching/version_graph.go

```diff
@@ -0,0 +1,223 @@
+		// Mutating the passed in graph is a no-no.
+		curDefault = proto.Clone(g.GetCurrentDefault()).(*taskqueuepb.VersionIdNode)
+		curNode := curDefault
+		for curDepth < maxDepth {
+			if curNode.GetPreviousIncompatible() == nil {
+				break
+			}
+			curNode = curNode.GetPreviousIncompatible()
+			curDepth++
+		}
+		if curNode != nil {
+			curNode.PreviousIncompatible = nil
+		}
+	}
+	return &workflowservice.GetWorkerBuildIdOrderingResponse{
+		CurrentDefault:   curDefault,
+		CompatibleLeaves: g.GetCompatibleLeaves(),
+	}
+}
+
+// Given an existing graph and an update request, update the graph appropriately.
+//
+// See the API docs for more detail. In short, the graph looks like one long line of default versions, each of which
+// is incompatible with the previous, optionally with branching compatibility branches. Like so:
+//
+// ─┬─1.0───2.0─┬─3.0───4.0
+//  │           ├─3.1
+//  │           └─3.2
+//  ├─1.1
+//  ├─1.2
+//  └─1.3
+//
+// In the above graph, 4.0 is the current default, and [1.3, 3.2] is the set of current compatible leaves. Links
+// going left are incompatible relationships, and links going up are compatible relationships.
+//
+// A request may:
+// 1. Add a new version to the graph, as a default version
+// 2. Add a new version to the graph, compatible with some existing version.
+// 3. Add a new version to the graph, compatible with some existing version and as the new default.
+// 4. Unset a version as a default. It will be dropped and its previous incompatible version becomes default.
+// 5. Unset a version as a compatible. It will be dropped and its previous compatible version will become the new
+//    compatible leaf for that branch.
+func UpdateVersionsGraph(
+	existingData *persistence.VersioningData,
+	req *workflowservice.UpdateWorkerBuildIdOrderingRequest,
+) error {
+	if req.GetVersionId().GetWorkerBuildId() == "" {
+		return serviceerror.NewInvalidArgument(
+			"request to update worker build id ordering is missing a valid version identifier")
+	}
+	// If the version is to become the new default, add it to the list of current defaults, possibly replacing
+	// the currently set one.
+	if req.GetBecomeDefault() {
+		curDefault := existingData.GetCurrentDefault()
+		if curDefault != nil {
+			// If the current default is going to be the previous compat version with the one we're adding,
+			// then we need to skip over it when setting the previous *incompatible* version.
+			if req.GetPreviousCompatible().GetWorkerBuildId() == curDefault.GetVersion().GetWorkerBuildId() {
+				existingData.CurrentDefault = &taskqueuepb.VersionIdNode{
+					Version:              req.VersionId,
+					PreviousCompatible:   curDefault,
+					PreviousIncompatible: curDefault.PreviousIncompatible,
+				}
+			} else {
+				// Otherwise, set the previous incompatible version to the current default.
+				existingData.CurrentDefault = &taskqueuepb.VersionIdNode{
+					Version:              req.VersionId,
+					PreviousCompatible:   nil,
```

Comment:
should this be req.GetPreviousCompatible()?

Reply:
Nope. In fact, if it's set at all that should really just be an error. Added comment explaining and made it work that way.